sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -1552,17 +1552,15 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
   override def visitTypeConstructor(ctx: TypeConstructorContext): Literal = withOrigin(ctx) {
     val value = string(ctx.STRING)
     val valueType = ctx.identifier.getText.toUpperCase(Locale.ROOT)
-    def toLiteral[T](f: UTF8String => Option[T], t: DataType): Literal = {
-      f(UTF8String.fromString(value)).map(Literal(_, t)).getOrElse {
+    def toLiteral(t: DataType): Literal = {
+      Try {Literal.fromString(value, t)}.getOrElse {
         throw new ParseException(s"Cannot parse the $valueType value: $value", ctx)
       }
     }
     try {
       valueType match {
-        case "DATE" => toLiteral(stringToDate, DateType)
-        case "TIMESTAMP" =>
-          val timeZone = getTimeZone(SQLConf.get.sessionLocalTimeZone)
-          toLiteral(stringToTimestamp(_, timeZone), TimestampType)
+        case "DATE" => toLiteral(DateType)
```

Comment:
can we just do `Cast(Literal(str), DateType)`? Or `Literal.create(Cast(...).eval, DateType)`

Reply:
I did that in another PR but @hvanhovell doesn't like that.

Reply:
Looks the current way is cleaner (see https://github.com/apache/spark/pull/23541/commits/016c696324a8203c7aaee92ea152796dc98139bb). I'm not sure where he said that in the PR though.